### PR TITLE
Refactor sort detectors to be more consistent

### PR DESCRIPTION
**Pull request by mbj**

### CODE REVIEW

**Comment by mbj** [comment]
@dkubb, @snusnu This is ready for review. Note that this is in fact batching up 3 PRs, you can mentally look as each `Change` commit to be the end of a separate PR.

File: src/SourceConstraints.hs [hunk 1]
```diff
@@ -115,31 +119,29 @@ warnings dynFlags (L sourceSpan node) =
         neverQualify
 
     descend :: Data a => a -> WarningMessages
-    descend =
-      unionManyBags . gmapQ
-        (descend `ext2Q` warnings dynFlags)
+    descend = unionManyBags . gmapQ (descend `ext2Q` warnings context)
```
Comment (snusnu) [inline src/SourceConstraints.hs]: @mbj This hunk is unrelated to the `Change` and should be extracted to a prefixed `Refactor` (or formatting `Fix`)

File: src/SourceConstraints.hs [hunk 2]
```diff
@@ -122,8 +122,16 @@ warnings context@Context{..} (L sourceSpan node) =
     descend :: Data a => a -> WarningMessages
     descend = unionManyBags . gmapQ (descend `ext2Q` warnings context)
 
-locatedWarning :: Data a => Context -> a -> Maybe ErrMsg
-locatedWarning context = mkQ empty sortedImportStatement
+data IEClass = Module String | Type String | Operator String | Function String
```
Comment (snusnu) [inline src/SourceConstraints.hs]: @mbj I'm curious if we have any rules regarding the canonical source location for type definitions. Is it "ok" to have this "interleaved" here with function definitions?

File: src/SourceConstraints.hs [hunk 3]
```diff
@@ -228,15 +214,10 @@ sortedLocated name context@Context{..} ordering nodes =
...
-
-    message :: Located a -> String
-    message expected =
-        "Unsorted " <> name <> ", expected: " <> render context expected
+        (text $ "Unsorted " <> name <> ", expected: " <> render context expectedNode)
 
     isViolation :: ((Located a, b), (Located a, b)) -> Bool
-    isViolation ((_actualNode, item), (_expectedNode, expected)) =
-      item /= expected
+    isViolation ((_actualNode, item), (_expectedNode, expected)) = item /= expected
```
Comment (snusnu) [inline src/SourceConstraints.hs]: @mbj This hunk is unrelated to the `Change`. What about formatting it this way right when the function is introduced in the previous commit?

File: src/SourceConstraints.hs (hunk 2 above)
Comment (mbj) [reply src/SourceConstraints.hs]: Its only used here, and its not exported. So I added it just above the only usage side.

File: src/SourceConstraints.hs (hunk 3 above)
Comment (mbj) [reply src/SourceConstraints.hs]: worse it violates 79chars, I could just kill this diff.

**Comment by mbj** [comment]
@dkubb, @snusnu This is ready for the next pass.

File: src/SourceConstraints.hs (hunk 2 above)
Comment (snusnu) [reply src/SourceConstraints.hs]: That seems fair to me.
